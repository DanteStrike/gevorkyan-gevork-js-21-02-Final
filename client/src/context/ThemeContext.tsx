import React from 'react';
import {DomUtils} from '../utils';

export enum Theme {
  LIGHT = `light`,
  DARK = `dark`,
  DEFAULT = `light`,
}

export interface IThemeContext {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}
interface IThemeProviderProps {
  children: React.ReactNode;
  storageKey: string;
}
interface IThemeProviderState {
  theme: Theme;
}

const ThemeContext = React.createContext<IThemeContext>({
  theme: Theme.LIGHT,
  setTheme: () => {},
});

class ThemeProvider extends React.PureComponent<IThemeProviderProps, IThemeProviderState> {
  constructor(props: IThemeProviderProps) {
    super(props);

    this.state = {
      theme: Theme.DEFAULT,
    };

    this.handleThemeChange = this.handleThemeChange.bind(this);
  }

  componentDidMount(): void {
    const {storageKey} = this.props;
    const storedTheme = (localStorage.getItem(storageKey) as Theme) || Theme.DEFAULT;

    this.setState({theme: storedTheme}, () => {
      DomUtils.addBodyClass(`theme--${storedTheme}`);
    });
  }

  handleThemeChange(value: Theme) {
    const {storageKey} = this.props;
    const {theme} = this.state;
    DomUtils.removeBodyClass(`theme--${theme}`);

    this.setState({theme: value}, () => {
      DomUtils.addBodyClass(`theme--${value}`);
    });
    localStorage.setItem(storageKey, value);
  }

  render() {
    const {theme} = this.state;
    const {children} = this.props;

    return <ThemeContext.Provider value={{theme, setTheme: this.handleThemeChange}}>{children}</ThemeContext.Provider>;
  }
}

export {ThemeContext, ThemeProvider};
