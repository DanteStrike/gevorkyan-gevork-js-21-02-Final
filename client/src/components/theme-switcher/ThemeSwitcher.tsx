import React, {useContext} from 'react';
import {Switch} from 'antd';
import './ThemeSwithcer.scss';
import {Theme as ContextThemes, ThemeContext} from '../../context/ThemeContext';
import useAppTranslation from '../../hooks/use-app-translation';

function ThemeSwitcher() {
  const themeContext = useContext(ThemeContext);
  const Theme = ContextThemes;
  const {t} = useAppTranslation(`theme`);

  const handleThemeChange = (checked: boolean) => {
    const {setTheme} = themeContext;

    if (checked) {
      setTheme(Theme.DARK);
    } else {
      setTheme(Theme.LIGHT);
    }
  };

  return (
    <Switch
      className="theme-switcher"
      checkedChildren={t(`dark`)}
      unCheckedChildren={t(`light`)}
      onChange={handleThemeChange}
      checked={themeContext.theme === Theme.DARK}
    />
  );
}

export default ThemeSwitcher;
