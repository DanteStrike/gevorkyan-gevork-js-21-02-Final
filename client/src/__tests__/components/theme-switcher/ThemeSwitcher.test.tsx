import React from 'react';
import {mount, shallow} from 'enzyme';
import {Theme, ThemeContext} from '../../../context/ThemeContext';
import ThemeSwitcher from '../../../components/theme-switcher/ThemeSwitcher';

describe(`Component ThemeSwitcher should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(<ThemeSwitcher />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component ThemeSwitcher should work correctly`, () => {
  it(`snapshot default test`, () => {
    const mockContext: any = {
      theme: Theme.LIGHT,
      setTheme: jest.fn((val) => {
        mockContext.theme = val;
      }),
    };

    const wrapper = mount(
      <ThemeContext.Provider value={mockContext}>
        <ThemeSwitcher />
      </ThemeContext.Provider>
    );
    const component = wrapper.find(`Switch`).at(0);
    expect(component.prop(`checked`)).toBe(false);
    component.simulate(`click`, true);
    expect(mockContext.setTheme).toHaveBeenLastCalledWith(Theme.DARK);

    const wrapperDark = mount(
      <ThemeContext.Provider value={mockContext}>
        <ThemeSwitcher />
      </ThemeContext.Provider>
    );
    const componentDark = wrapperDark.find(`Switch`).at(0);
    expect(componentDark.prop(`checked`)).toBe(true);
    componentDark.simulate(`click`, false);
    expect(mockContext.setTheme).toHaveBeenLastCalledWith(Theme.LIGHT);

    const wrapperLight = mount(
      <ThemeContext.Provider value={mockContext}>
        <ThemeSwitcher />
      </ThemeContext.Provider>
    );
    const componentLight = wrapperLight.find(`Switch`).at(0);
    expect(componentLight.prop(`checked`)).toBe(false);
  });
});
