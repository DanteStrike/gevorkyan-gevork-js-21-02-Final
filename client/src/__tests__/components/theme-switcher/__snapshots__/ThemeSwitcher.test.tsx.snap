// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ThemeSwitcher should render correctly snapshot default test 1`] = `
<Switch
  checked={false}
  checkedChildren="Dark theme"
  className="theme-switcher"
  onChange={[Function]}
  unCheckedChildren="Light theme"
/>
`;
